File: vbox-app/src/main/java/com/kedzie/vbox/machine/LogFragment.java
package com.kedzie.vbox.machine;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.kedzie.vbox.R;
import com.kedzie.vbox.api.IMachine;
import com.kedzie.vbox.app.BundleBuilder;
import com.kedzie.vbox.task.ActionBarTask;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * 
 * @author Marek Kędzierski
 * @apiviz.stereotype fragment
 */
public class LogFragment extends Fragment {
    private static final String TAG = "LogFragment";
	private static final int MAX_LOG_SIZE=409600; //400 Kbps
	
	class LoadLogTask extends ActionBarTask<IMachine, String> {
		public LoadLogTask() {
			super((AppCompatActivity)getActivity(), null);
		}

		@Override 
		protected String work(IMachine... m) throws Exception {
		    try {
		        return new String(m[0].readLog(0, 0, MAX_LOG_SIZE));
		    } catch(Exception e) {
		        return "";
		    }
		}

		@Override
		protected void onSuccess(String result) {
			if(result.length()==MAX_LOG_SIZE)	
			    Log.w(TAG,"Didn't get entire log file.  Log size: " + result.length());
			_logText.setText(_log=result);
		}
	}

	@BindView(R.id.logText)
	 TextView _logText;
	private IMachine _machine;
	private String _log;

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		_machine = BundleBuilder.getProxy(getArguments(), IMachine.BUNDLE, IMachine.class);
		if(savedInstanceState!=null)
			_log = savedInstanceState.getString("log");
	}
	
	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
		return inflater.inflate(R.layout.machine_log, null);
	}

	@Override
	public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
		super.onViewCreated(view, savedInstanceState);
		ButterKnife.bind(this, view);
	}

	@Override
	public void onStart() {
		super.onStart();
		if(_log!=null) 
			_logText.setText(_log);
		else
			new LoadLogTask().execute(_machine);
	}

	@Override
	public void onSaveInstanceState(Bundle outState) {
		outState.putString("log", _log);
	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		switch(item.getItemId()) {
		case R.id.option_menu_refresh:
		    Log.i(TAG, "Refreshing...");
			new LoadLogTask().execute(_machine);
			return true;
		}
		return false;
	}
}

File: vbox-app/src/main/java/com/kedzie/vbox/machine/LogFragment.kt
package com.kedzie.vbox.machine

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.kedzie.vbox.R
import com.kedzie.vbox.api.IMachine
import com.kedzie.vbox.app.BundleBuilder
import kotlinx.android.synthetic.main.machine_log.*
import kotlinx.coroutines.*
import timber.log.Timber
import kotlin.coroutines.CoroutineContext

class LogFragment : Fragment(), CoroutineScope {

    private lateinit var job: Job

    override val coroutineContext: CoroutineContext
        get() = job + Dispatchers.Default

    private lateinit var _machine: IMachine
    private var _log: String? = null

    private fun loadLog() {
        launch {
            var text = ""
            try {
                text = String(_machine.readLog(0, 0, MAX_LOG_SIZE.toLong()))
            } catch (e: Exception) {
                Timber.w(e, "Error reading log")
            }

            withContext(Dispatchers.Main) {
                if (text.length == MAX_LOG_SIZE)
                    Timber.w( "Didn't get entire log file.  Log size: %d", text.length)
                _log = text
                logText!!.text = _log
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        _machine = BundleBuilder.getProxy(arguments, IMachine.BUNDLE, IMachine::class.java)
        if (savedInstanceState != null)
            _log = savedInstanceState.getString("log")
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val swipeLayout = SwipeRefreshLayout(activity!!)
        swipeLayout.setOnRefreshListener { loadLog() }
        inflater.inflate(R.layout.machine_log, swipeLayout)
        return swipeLayout
    }

    override fun onStart() {
        super.onStart()
        job = Job()
        if (_log != null)
            logText!!.text = _log
        else
            loadLog()
    }

    override fun onStop() {
        job.cancel()
        super.onStop()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putString("log", _log)
    }

    companion object {
        private const val MAX_LOG_SIZE = 409600 //400 Kbps
    }
}
